일별 카드 개수 조회 API 구현

Park-SooYeon opened the pull request:
# 일별 카드 개수 조회 API 구현

## 📎 해당 이슈

이슈 링크 (#110)

## 🛠 구현 내용 

- 일별 카드 개수 조회 API 구현
- API 및 service 함수를 테스트 하는 테스트 코드 작성

## 🙋‍ 리뷰어 참고 사항

SQL문에서 from 절에 서브 쿼리를 사용하는 것은 `인라인뷰`라고 부른다더군요.
DB에서 내부적으로 쿼리를 실행할 때, 일반적으로 인라인 뷰는 결국 join문으로 변경되서 실행이 된다고 합니다.
고로 인라인뷰를 썼을 때와 그냥 inner join을 사용했을 때의 성능차이가 나지 않는다는 결론이 나왔습니다.
실제로 explain 명령을 사용해서 실제 실행되는 쿼리를 확인해 보았을 때도 동일한 쿼리결과가 나왔었습니다.
그래서 join문을 이용해서 일별 카드 개수 조회 쿼리를 작성하였습니다:D

Park-SooYeon committed:
[FEAT] 일별 카드 개수를 조회하는 서비스 함수 추가

- 일별 카드 개수를 조회하는 getCardCountByPeriod 함수 추가

youngxpepp committed:
[CHORE] ESLint 에러 해결

- no-unused-vars가 어쩔 수 없이 생기는 경우가 있으므로 warn으로 설정
- 대부분의 no-unused-vars 해결
- console.log 제거
- 테스트 실패 시 테스트가 끝나지 않는 문제로 인해 Application.initialize()의 try-catch 제거

youngxpepp committed:
[FEAT] Passport NaverStrategy 환경변수 기본값 설정

- 테스트 시 환경변수가 설정되어 있지 않으면 에러 발생

youngxpepp committed:
[FIX] 테스트 종료 시 데이터베이스 커넥션을 함께 종료

- 테스트가 끝나도 특정 비동기 동작이 수행 중 -> 데이터베이스 커넥션
- 테스트 afterAll에 커넥션 종료 명령 추가

Park-SooYeon committed:
[FEAT] 일별 카드 개수 조회 시, boardIds 조건 추가

- 로그인 한 유저가 참여 중인 보드의 카드 개수만 조회해야 되므로 boardIds 조건 추가
- 카드 개수 alias명 변경

Park-SooYeon committed:
Merge branch 'develop/web' of https://github.com/boostcamp-2020/Project04-B-Whale into feature/get-card-count

youngxpepp committed:
[FEAT] SQLite 데이터베이스 저장 공간 지정 가능

- SQLite를 파일 시스템 또는 메모리 어디에서 관리할지 정할 수 있음

youngxpepp committed:
[FEAT] develop/web의 서버 CI 설정

- ESLint로 코드 컨벤션 검사
- Babel 빌드
- MySQL 컨테이너 환경에서의 테스트

Park-SooYeon committed:
Merge branch 'develop/web' of https://github.com/boostcamp-2020/Project04-B-Whale into feature/get-card-count

Park-SooYeon committed:
[TEST] CardService의 getCardCountByPeriod 테스트 코드 추가

- 주어진 기간동안 카드 개수가 정상적으로 조회되는지 테스트하는 테스트 코드 추가

Park-SooYeon committed:
[FIX] 테스트 코드 실패 오류 수정

- getCardCountByPeriod의 매개변수를 객체 형식으로 변경
- 쿼리 결과의 count 컬럼의 데이터 타입이 string으로 나오는 문제 수정을 위해 bigNumberString 옵션 추가

Park-SooYeon committed:
[FEAT] card api 사용을 위한 card router 추가

- query string의 값을 변경해주는 queryParser 함수 추가
- card router 추가

Park-SooYeon committed:
[REFACTOR] boardIds를 가져오는 코드 리팩토링

- getCardCountByPeriod함수의 statement를 prepare statement 방식으로 변경
- boardId만을 배열로 받아오는 getBoardIdsByUserId 함수 추가

Park-SooYeon committed:
[TEST] GET /api/card/count 테스트 코드 추가

##### 제목은 최대 50 글자까지만 입력 ############## -> |


# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
#     타입과 제목은 영어로 작성하고 본문은 한글로 작성한다.
#     관련된 이슈번호는 제목 맨 뒤에 추가한다. ex. (#1)
# ------------------

dh00n commented on server/src/common/util/queryParser.js:
```diff
@@ -0,0 +1,18 @@
+    queryString.split(separator).forEach((ele) => {
+        const keyValue = ele.split(delimiter);
+        const key = keyValue[0];
+        const value = keyValue[1];
```
```
const [key, value] = ele.split(delimiter);
```
로 작성해도 보기 좋을 것 같아유!!!!!

Park-SooYeon replied:
헉 그러네요 저쪽은 왜 저렇게 할 생각을 못했었을까요ㅋㅋㅋ
commit b79c971
수정했습니다!

dh00n reviewed (dismissed):
👍👍👍

Park-SooYeon committed:
[REFACTOR] 리뷰에 따라 코드 수정

- key, value를 구조 분해 할당 구문으로 수정

youngxpepp commented on server/src/service/CardService.js:
```diff
@@ -10,4 +11,22 @@ export class CardService extends BaseService {
+            .createQueryBuilder('card')
+            .select(`date_format(card.due_date, '%Y-%m-%d')`, 'dueDate')
+            .addSelect('count(1)', 'count')
+            .innerJoin('card.list', 'list', 'list.board_id IN(:boardIds)', { boardIds })
```
```suggestion
            .innerJoin('card.list', 'list', 'list.board_id IN(:...boardIds)', { boardIds })
```
https://typeorm.io/#/select-query-builder/using-parameters-to-escape-data
위 문서 보면 `...` 을 붙여줘야 하는걸로 나오는데, 확인해보셨나요?

Park-SooYeon replied:
cardRouter 테스트 코드 보면은 내가 만든 보드와 초대 받은 보드를 give로 하나씩 생성하고 테스트를 진행하고 있습니다.
두개의 보드가 있으니 boardIds가 [boardId1, boardId2]로 나올 것으로 예상했고, 테스트가 정상 통과가 되서 ...을 안써도 문제는 없는 듯해서 그냥 사용했는데 공식문서에서 안내해주는대로 사용하는게 좋을거 같네요.
수정하겠습니다:)

youngxpepp commented on server/src/service/BoardService.js:
```diff
@@ -40,6 +40,20 @@ export class BoardService extends BaseService {
+            .createQueryBuilder('board')
+            .select('board.id', 'id')
+            .leftJoin('board.invitations', 'invitation')
+            .where(`board.creator_id=:userId or user_id=:userId`, { userId })
```
```suggestion
            .where(`board.creator_id=:userId or invitation.user_id=:userId`, { userId })
```
invitation에 대한 조건이 빠진듯 해요

Park-SooYeon replied:
commit 600a091
위의 리뷰 내용과 해당 리뷰 내용에 따라 코드 수정하였습니다:)

youngxpepp commented on server/src/router/CardRouter.js:
```diff
@@ -0,0 +1,29 @@
+        const { id: userId } = req.user;
+        const { startdate: startDate, enddate: endDate, member } = queryParser(q);
+
+        const boardIds = await boardService.getBoardIdsByUserId(userId);
+
+        const config = { startDate, endDate, boardIds };
+        if (member === 'me') {
+            config.userId = userId;
+        }
+
+        const cardCounts = await cardService.getCardCountByPeriod(config);
```
하나의 트랜잭션에서 처리하는건 어떨까요?
나눈 이유가 궁금합니다!

Park-SooYeon replied:
확장성을 고려해서 member가 me가 아닌 nickname일때에는 해당 nickname을 가지는 유저의 id값을 가져와서 분기해줄 수 있을 것 같다는 생각에 controller에서 처리를 했습니다!
만약, nickname을 통한 검색 기능이 필요하다면 이후에 해당 코드를 아래와 같이 사용하여 controller에서 처리를 해준다면 service함수에는 수정되지 않아도 될 것 같다는 생각이 들어서 이렇게 작성했습니다!
```javascript
if (member === 'me') {
    config.userId = userId;
} else (member !== '') {
    config.userId = await userService.getUserIdByName(member);
}
```

youngxpepp replied:
확장성을 생각하셨군요👍
17번 라인이 실행되고 나서 트랜잭션이 끝나게 되는 순간 커넥션이 다른 요청에게 할당되면 어떡하죠?
컨트롤러 쪽 말고 서비스에서 분리하는건 어때요? 하나의 트랜잭션으로 묶을 수 있고, 확장성도 고려할 수 있을 것 같다는 생각이 듭니다ㅎㅎ

Park-SooYeon replied:
흠..확실히 하나의 트랜잭션에 DB select 하는 부분이 묶이지 않으면은 중간에 커넥션이 다른 요청에게 할당될수도 있겠네요🤔
서비스 쪽에서 분기를 해주게 되면은 서비스 함수의 매개변수로 현재 로그인된 userId를 필수로 받아야되는데 member가 me가 아닐때에는 해당 매개변수를 사용하는 일이 없으니 계속 받아올 필요가 있겠나라는 생각이 들어 controller에서 처리를 해줬는데 트랜잭션에 대해서는 생각을 못해봤네요.
한번 수정해보겠습니다!

Park-SooYeon replied:
commit aaee4e8
수정해보았습니다!
코드를 수정하다보니 트랜잭션 고려뿐만아니라 member값에 따른 분기는 비즈니스 로직으로 볼 수 있어 서비스가 해야하는 일이 맞았겠구나라는 생각이 들었네요..(아직 컨트롤러와 서비스의 역할이 잘 구분 하지 못해서 그런가봅니다😂)
그리고 컨트롤러는 단지 입력값 검증과 필요한 서비스 함수를 호출 하는 역할만 해야겠다고 생각하고 구현하려고 했으나.. validation을 어떤 방식으로 체크하는 것이 좋을지 고민이 되더라구요🤔
내일 validation을 어떤 방식으로 수행할지 얘기 나눠보면 좋을 것 같아요:)

youngxpepp replied:
늦은 시간까지 고생하셨습니다🌝
validation 수행 방식은 내일 같이 얘기해봐요!

youngxpepp commented on server/src/service/CardService.js:
```diff
@@ -10,4 +11,22 @@ export class CardService extends BaseService {
+            .groupBy(`date_format(card.due_date, '%Y-%m-%d')`);
+
+        if (userId) {
+            query = query.innerJoin('card.members', 'member', 'member.user_id=:userId', { userId });
```
여기가 me 조건인거죠?

Park-SooYeon replied:
넵 me 조건에 의해 userId가 매개변수로 넘어왔을 때에만 해당 쿼리를 추가하도록 조건을 추가한것입니다.

youngxpepp reviewed (changes requested):
고생하셨어요! 쿼리 설명 감사합니다👏

Park-SooYeon committed:
[REFACTOR] 리뷰에 따라 코드 수정

- 공식문서 사용법에 따라 쿼리 배열 파라미터에 ... 추가
- 쿼리 가독성을 위해 table alias 추가

youngxpepp commented on server/src/common/config/database/ConnectionOptionGenerator.js:
```diff
@@ -21,6 +21,7 @@ class ConnectionOptionGenerator {
                 connectionOption.url = this.databaseEnv.getDatabaseUrl();
                 connectionOption.connectTimeout = 3000;
                 // connectionOption.acquireTimeout = 5000;
+                connectionOption.bigNumberStrings = false;
```
아깐 제대로 못 봤었는데 이건 어떠한 이유로 추가하신건지 궁금합니당😃

Park-SooYeon replied:
일단, 집계함수의 결과를 number type으로 받아들이기 위해서 위 속성을 사용했습니다.
count와 sum 같은 집계함수를 사용할 경우에는 쿼리문 결과를 가져올 때, .getMany()를 사용할 수 없고, .getRawMany()를 사용할 수 있고, 집계함수의 결과를 number type이 아닌 아래와 같이 string type으로 반환해 주더라구요...
``` json
{
    "dueDate": "2020-07-01",
    "count": "2"
}
```
제 생각에 집계함수를 사용할 경우에는 엄청 큰 값이 집계함수의 결과 값으로 주어질 수 있기때문에 typeORM이 기본적으로 집계함수의 결과를 string으로 반환해주는 것 같았습니다.🤔
카드 개수가 막 오버플로우가 발생할 정도로 엄청 큰 숫자로 반환되지는 않을 것 같고, 다른 부분에서도 엄청 큰 숫자를 쿼리문의 결과를 받아들일 경우도 없을거라고 생각되고, count의 값을 number type으로 일일히 바꿔주는 것보다 해당 속성을 추가하는 것이 더 효율적일 것이라고 생각해서 추가하게 되었습니다.
사실 response data의 count 값의 type이 string이여도 상관이 없었을까요.

youngxpepp replied:
꼼꼼함에 감탄하고 갑니다! 이런 설정이 있는줄 몰랐네요. 배워 갑니다👏

Park-SooYeon committed:
[REFACTOR] 리뷰에 따라 서비스함수 로직 변경

- member 값에 따른 분기를 서비스 함수에서 수행하도록 변경
- 서비스 함수 변경에 따른 테스트 코드 변경

youngxpepp reviewed (dismissed):
섬세하게 코딩하시는 모습이 인상적입니다! 배워갑니당!

Park-SooYeon committed:
[FIX] 'develop/web' pull confilct 해결

- remote 저장소의 develop/web branch pull할 때, confilct 해결

Park-SooYeon committed:
Merge branch 'develop/web' of https://github.com/boostcamp-2020/Project04-B-Whale into feature/get-card-count

Park-SooYeon committed:
[FEAT] validation을 위한 함수 및 DTO 추가

- validation에 필요한 validator 함수 추가
- 일별 카드 개수 조회 API에서 validation에 사용할 DTO 추가
- validation error 객체 추가

Park-SooYeon committed:
[FEAT] 일별 카드 개수 조회 API validation 체크 로직 추가

- 일별 카드 개수 조회 API의 controller 부분에 validation 체크 로직 추가

Park-SooYeon committed:
[TEST] 일별 카드 개주 조회 API 테스트 케이스 추가

validation 체크 로직 추가에 따라 테스트 케이스 추가

youngxpepp commented on server/src/common/util/validator.js:
```diff
@@ -0,0 +1,11 @@
+
+export const validator = async (DtoClass, Object, groups) => {
+    const classObject = plainToClass(DtoClass, Object);
+    const [error] = await validate(classObject, { groups });
+    if (error) {
+        throw new ValidationError('validation failed');
+    }
```
```suggestion
    await validateOrReject(classObject, { groups });
```
이렇게 하면 에러를 따로 던져주지 않아도 될 것 같아요✌️
ValidationError는 이미 'class-validator'에 정의가 되어 있어서 미들웨어인 errorHandler에서 처리하고 있습니다!
https://github.com/boostcamp-2020/Project04-B-Whale/blob/2a890211ab822d47472be513469b7538ad71483e/server/src/common/middleware/errorHandler.js#L13-L22

Park-SooYeon replied:
validator함수가 미들웨어로 처리되는게 아니여서 그런지 validateOrReject를 사용하면 에러 핸들링이 되지 않는다는 오류가 뜨더라구요..
그래서 validateOrReject에 의해 생성되는 에러를 핸들링 하려면 코드를 아래와 같이 작성해야하는데 try-catch에서 아무런 작업없이 에러를 던지기만하는 것은 린트 설정으로 막혀져있더라구요.
``` javascript
try {
    // ... 생략
     await validateOrReject(classObject, { groups });
} catch (e) {
    throw e;
}
```
차라리 린트 설정을 수정하는게 나았을까요?🤔

youngxpepp replied:
흠... 왜 처리가 안될까요
errorHandler 부분 디버깅 해주실 수 있나요?

Park-SooYeon replied:
다시 해봤는데 아무런 문제가 없네요..
코드 작성하면서 이것저것 같이 테스트했었는데 거기서 문제가 있었었나봅니다..😂
commit 90e7527 에 리뷰 반영했습니다!

youngxpepp reviewed (changes requested):
진짜 고생하십니다😂
Validation 부분 한번 봐주실래요?

Park-SooYeon committed:
[FIX] 리뷰에 따라 코드 수정

- validate함수 대신 validateOrReject 함수 사용으로 변경
- validateOrReject 함수 사용에 따라 ValidationError 클래스 제거

youngxpepp reviewed (approved):
고생하셨습니다!

dh00n reviewed (approved):
👍👍👍👍👍

youngxpepp committed:
Merge branch 'develop/web' into feature/get-card-count
